ls1intum/Artemis
# [Feature] Grading statistics

### Checklist

### Motivation and Context



Add statistics to the grading page for instructors.

### Description


##### Charts & Statistics
- Added a Server REST call to generate the statistics
- Added charts for each test case showing the percentage of students that passed the test
- Added charts for each sca category showing the amount of issues detected across students
- Added charts area next to the datatable
  - Created a new chart component as a wrapper around the chartjs directive
  - Added one chart for each table showing different metrics regarding the test cases / sca categories.

##### Other
- Made some implovements to the overall layout of the grading page.
- Added '?' icons for the most important columns of the tables.
- Included the colors from the charts in the datatable.

### Steps for Testing


1. Create a programming exercise
2. Particpate in the exercise with multiple accounts and different solutions
    - Solve different test cases
    - Solve / Introduce different sca issues
3. Open the grading page and check the statistics
     - Tip: All charts have tooltips
4. Edit the test case and category configuration
    - The charts should update according to the changes
6. Save your changes and trigger a rebuild
7. See the updated results as the students

### Test Coverage


##### Server
- **ProgrammingExerciseGradingService**: 93% stmts
- **ProgrammingExerciseGradingStatisticsDTR**: 86% stmts

##### Client
- **programming-exercise-configure-grading.component.ts**: 82.35% stmts
- **sca-category-distribution-chart.component.ts**: 89% stmts
- **test-case-distribution-chart.component.ts**: 100% stmts
- **test-case-passed-builds-chart.component.ts**: 100% stmts
- **category-issues-chart.component.ts**: 100% stmts

*(More elaborate tests and load tests will be added in a follow-up PR)*

### Screenshots


<img width="1440" alt="Bildschirmfoto 2020-10-24 um 17 57 46" src="https://user-images.githubusercontent.com/13920539/97086274-81f5c980-1622-11eb-9654-aee8872f2e28.png">
<img width="1438" alt="Bildschirmfoto 2020-10-24 um 17 57 57" src="https://user-images.githubusercontent.com/13920539/97086277-8621e700-1622-11eb-9031-b2e1cf14b59e.png">
